File: src/Python/250.CountUnivalueSubtrees.py

class Solution:
    
    def __init__(self):
        self.ans = 0

    def dfs(self, root, pre):
        if root is None:
            return True
        flag1 = self.dfs(root.left, root.val)
        flag2 = self.dfs(root.right, root.val)
        if flag1 == True and flag2 == True:
            ans += 1
        return root.val == pre and flag1 and flag2


    def countUnivalSubtrees(self, root):
        if root is None:
            return 0
        self.ans = 0
        self.dfs(root, root.val)
        return self.ans

    
if __name__ == "__main__":
    solution = Solution()

File: src/Java/250.CountUnivalueSubtrees.java

import java.util.List;
import java.util.TreeSet;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Queue;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Random;


class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) {
        val = x;
    } 
}


class Solution {
    boolean dfs(TreeNode root, int pre, Integer ans) {
        if (root == null) 
            return true;
        boolean flag1 = dfs(root.left, root.val, ans);
        boolean flag2 = dfs(root.right, root.val, ans);
        if (flag1 && flag2) 
            ans++;
        return (root.val == pre) && flag1 && flag2;
    }

    public int countUnivalSubtrees(TreeNode root) {
        if (root == null) 
            return 0;
        Integer ans = 0;
        dfs(root, root.val, ans);
        return ans;
    }
}
